### PR TITLE
Store error backtraces on discrete executions

### LINKED ISSUE #1162: Storing backtrace in database?
Loving GoodJob so far, thanks so much for this!

I was wondering if there's an option to store the backtrace (or at least a few lines of it) in the good_job_executions table. It'd be handy to have that information in the database and UI in order to determine the specific cause of a failure, and so that once I fix a bug, I can easily identify all jobs caused by that same bug and retry them. The `error` column often isn't specific enough for me to pinpoint this.

Am I correct in my read that the best way to get the backtrace right now is to set `on_thread_error` and report the error to Rails, which then would have it printed to the logs and/or sent to a tool like Sentry?

Thanks!

### PR DESCRIPTION
**Pull request by Intrepidd**
Closes #1162 

Screenshot 2024-04-18 at 5 43 32 PM

### CODE REVIEW

File: app/models/good_job/execution.rb [hunk 1]
```diff
@@ -455,12 +459,14 @@ def perform
...
         if job_error
           error_string = self.class.format_error(job_error)
+          error_backtrace = self.class.format_backtrace(job_error.backtrace)
 
           job_attributes[:error] = error_string
           job_attributes[:error_event] = result.error_event if self.class.error_event_migrated?
           if discrete_execution
             discrete_execution.error = error_string
             discrete_execution.error_event = result.error_event
+            discrete_execution.error_backtrace = error_backtrace if discrete_execution.class.backtrace_migrated?
```
Comment (bensheldon) [inline app/models/good_job/execution.rb:469]: Just to save a little bit of unnecessary processing, let's only format the backtrace if it's been migrated:

```suggestion
            if discrete_execution.class.backtrace_migrated?
              error_backtrace = self.class.format_backtrace(job_error.backtrace)
              discrete_execution.error_backtrace = error_backtrace
            end
```

**Review by bensheldon — COMMENTED** [review]
This is awesome and super clean 👍🏻 

Would you want to also render this out into the Dashboard view too in this PR? 

Otherwise I have one tiny suggestion, and the test that fails is a flake, so this looks awesome 🎉

**Comment by Intrepidd** [comment]
Thanks for the swift review ! 

> Would you want to also render this out into the Dashboard view too in this PR?

You can already kinda see it in the details of an execution in the `_good_job_execution` key.

I figured this is good enough for this PR and could be released as-is to provide some value, but if you have pointers on how you would like it to be displayed as a first class citizen, I'm glad to make a PR for it

**Comment by bensheldon** [comment]
No sweat! I was poking around to form an opinion and just did it 😁 

Thank you for working on this. It's great!

**Comment by Earlopain** [comment]
This is pretty neat, thanks for working on this👍

If I am understanding this corretly, the cleaned backtrace will be saved and only the first frame displayed, is that correct? There have been several occasions where my application backtrace didn't include the full context, for one or another reason.

Would it make sense to store the full backtrace and format only on display? I'm thinking of something like how Rails displays errors in development.

**Pull request merged**